src/navigation-bar.tsx

```diff
@@ -129,11 +242,1008 @@ type MainNavigationBarPropsBase = {
     withBorder?: boolean;
     burgerMenuExtra?: React.ReactNode;
     large?: boolean;
+    desktopSmallMenu?: boolean;
+};
+
+type MainNavigationBarMenuStatus = 'opening' | 'opened' | 'closing' | 'closed';
+type MainNavigationBarMenuAction = 'open' | 'finishOpen' | 'close' | 'finishClose';
+
+const mainNavigationDesktopMenuTranstions: Record<
+    MainNavigationBarMenuStatus,
+    Partial<Record<MainNavigationBarMenuAction, MainNavigationBarMenuStatus>>
+> = {
+    opening: {
+        close: 'closing',
+        finishOpen: 'opened',
+    },
+    opened: {
+        close: 'closing',
+    },
+    closing: {
+        // If a section was opened while the menu was closing, the menu should be considered as
+        // already open. This is useful for example to avoid the new content's fade-in animation
+        open: 'opened',
+        finishClose: 'closed',
+    },
+    closed: {
+        open: 'opening',
+    },
+};
+
+const menuReducer = (state: MainNavigationBarMenuStatus, action: MainNavigationBarMenuAction) => {
+    return mainNavigationDesktopMenuTranstions[state][action] || state;
+};
+
+type MainNavigationBarDesktopMenuState = {
+    isMenuOpen: boolean;
+    openedSection: number;
+    menuHeight: string;
+    menuStatus: MainNavigationBarMenuStatus;
+    openSectionMenu: (index: number) => void;
+    closeMenu: () => void;
+    setMenuHeight: (height: string) => void;
+    setIsMenuHovered: (value: boolean) => void;
+    setFocusedItem: (item?: {column: number; index: number}) => void;
+    debouncedOpenSectionMenu: (index: number) => void;
+    cancelDebouncedOpenSectionMenu: (index: number) => void;
+};
+
+const MainNavigationBarDesktopMenuContext = React.createContext<MainNavigationBarDesktopMenuState>({
+    isMenuOpen: false,
+    openedSection: -1,
+    menuHeight: '0px',
+    menuStatus: 'closed',
+    openSectionMenu: () => {},
+    closeMenu: () => {},
+    setMenuHeight: () => {},
+    setIsMenuHovered: () => {},
+    setFocusedItem: () => {},
+    debouncedOpenSectionMenu: () => {},
+    cancelDebouncedOpenSectionMenu: () => {},
+});
+
+const MainNavigationBarDesktopMenuContextProvider = ({
```

Comment:
This context handles the entire logic of the menu and exposes required callbacks to update different things related to it